ThunderCloud/TSCAppDelegate.swift

```diff
@@ -0,0 +1,196 @@
+//
+//  TSCAppDelegate.swift
+//  ThunderCloud
+//
+//  Created by Simon Mitchell on 24/08/2017.
+//  Copyright © 2017 threesidedcube. All rights reserved.
+//
+
+import UIKit
+import UserNotifications
+import ThunderRequest
+
+@UIApplicationMain
+/// A root app delegate which sets up your window and push notifications e.t.c.
+open class TSCAppDelegate: UIResponder, UIApplicationDelegate, UNUserNotificationCenterDelegate {
+
+	/// The main window of the app
+	public var window: UIWindow?
+	
+	/// Whether to show push notifications when the app is in the foreground
+	public var foregroundNotificationOptions: UNNotificationPresentationOptions? = [.alert, .badge, .sound]
+	
+	open func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey : Any]? = nil) -> Bool {
+		
+		UNUserNotificationCenter.current().delegate = self
+		
+		window = UIWindow(frame: UIScreen.main.bounds)
+		window?.backgroundColor = .white
+		
+		window?.rootViewController = TSCAppViewController()
+		window?.makeKeyAndVisible()
+		
+		setupSharedUserAgent()
+		
+		DeveloperModeController.shared.installDeveloperMode(toWindow: window!, currentTheme: TSCTheme())
+		
+		// Register errors
+		TSCErrorRecoveryAttempter.registerOverrideDescription(
+			"Failed to load page".localised(with: "_STREAMINGPAGE_FAILED_TITLE"),
+			recoverySuggestion: "We were unable to find the page for this notification".localised(with: "_STREAMINGPAGE_FAILED_RECOVERYSUGGESTION"),
+			forDomain: "ThunderCloud.streamingError",
+			code: 1)
+				
+		return true
+	}
+	
+	//MARK: -
+	//MARK: - Push notifications
+	//MARK: -
+	
+	open func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
+		TSCStormNotificationHelper.registerPushToken(deviceToken)
+	}
+
+	open func handleNotificationResponse(for notification: UNNotification, response: UNNotificationResponse?, fromLaunch: Bool) -> Bool {
+		
+		// Make sure the action wasn't dismissing the notification
+		guard response?.actionIdentifier != UNNotificationDismissActionIdentifier else {
+			return false
+		}
+		
+		guard let payload = notification.request.content.userInfo["payload"] as? [AnyHashable : Any] else { return false }
+		guard let url = payload["url"] as? String else { return false }
+		
+		// Call handleNotificationResponse(for:response:fromLaunch) every second if we are presenting
+		if window?.rootViewController?.presentedViewController != nil {
```

Comment:
I suppose this could also be a guard too, so that it handles `return handleContentNotificationFor(cacheURL: url)` first. Keeps the golden path neat, as above :)